src/pybind/mgr/volumes/fs/stats_util.py

```diff
@@ -0,0 +1,328 @@
+        # LibCephFS.getxattr() can be made.
+        self.vc = vc
+
+        # Creating an RTimer instance in advance so that we can check if clone
+        # reporting has already been initiated by calling RTimer.is_alive().
+        self.update_task = RTimer(1, self._update_progress_bars)
+
+    def initiate_reporting(self):
+        if self.update_task.is_alive():
+            return
+
+        # progress event ID for ongoing clone jobs
+        self.on_pev_id: Optional[str] = str(uuid4())
+        # progress event ID for ongoing+pending clone jobs
+        self.onpen_pev_id: Optional[str] = str(uuid4())
+        self.show_onpen_bar = False
+
+        self.update_task = RTimer(1, self._update_progress_bars)
+        self.update_task.start()
+
+    def _get_clone_dst_info(self, fsh, ci, clone_entry, clone_index_path):
+        ce_path = os_path_join(clone_index_path, clone_entry)
+        # XXX: This may raise ObjectNotFound exception. As soon as cloning is
+        # finished, clone entry is deleted by cloner thread. This exception is
+        # handled in _get_info_for_all_clones().
+        ci.dst_path = fsh.readlink(ce_path, 4096).decode('utf-8')
+
+        ci.dst_grp_name, ci.dst_sv_name = \
+            resolve_grp_and_sv_names(self.vol_spec, ci.dst_path)
+
+    def _get_clone_src_info(self, fsh, ci):
+        with open_clone_sv_pair_in_vol(
+                self.vc, self.vol_spec, ci.volname, ci.dst_grp_name,
+                ci.dst_sv_name) as (dst_sv, src_sv, snap_name):
+            ci.src_grp_name = src_sv.group_name
+            ci.src_sv_name = src_sv.subvolname
+            ci.src_path = src_sv.snapshot_data_path(snap_name)
+
+    def _get_info_for_all_clones(self):
+        clones = []
+
+        volnames = list_volumes(self.vc.mgr)
+        for volname in volnames:
+            with open_volume_lockless(self.vc, volname) as fsh:
+                with open_clone_index(fsh, self.vol_spec) as clone_index:
+                    clone_index_path = clone_index.path
+                    clone_entries = clone_index.list_entries_by_ctime_order()
+
+            for ce in clone_entries:
+                ci = CloneInfo()
+                ci.volname = volname
+
+                try:
+                    self._get_clone_dst_info(fsh, ci, ce, clone_index_path)
+                    self._get_clone_src_info(fsh, ci)
+                    if ci.src_path is None or ci.dst_path is None:
+                        continue
+                # clone entry went missing, it was removed because cloning has
+                # finished.
+                except ObjectNotFound:
+                    continue
+
+                clones.append(ci)
+
+        return clones if clones else []
+
+    def _update_progress_bar_event(self, ev_id, ev_msg, ev_progress_fraction):
+        # in case this remote call on RTimer fails, its details won't printed,
+        # logged or would cause a crash. therefore, leaving some info for
```

Comment:
just being curious, what could be the possible crashes? I couldn't think of.

Reply:
Not expecting crash but whenever crash happened during early testing it wasn't easy to spot it since only a line was printed in MGR logs for this crash.